### PR TITLE
Add multi label classification task

### PR DESCRIPTION
**Pull request by Tyrest**
This PR adds the `MultiLabelClassificationTask` to `autolabel`. Currently tests the prompt generation and `MultiLabelClassificationTask.eval`. Below is an example prompt:

```

You are an expert at classifying tweets as neutral or one or more of the given emotions that best represent the mental state of the poster.
Your job is to correctly label the provided input example into one or more of the following categories:
neutral
anger
anticipation
disgust
fear
joy
love
optimism
pessimism
sadness
surprise
trust

You will return the answer as a comma separated list of labels sorted in alphabetical order. For example: "label1, label2, label3"

Some examples with their output answers are provided below:

Input: In my room 101 would go  Russell Howard,Tom Odell,Michael Buble!!! #pants #nogood 
Output: anger, disgust, pessimism, sadness

Input: @FullTimeDEVILS Memphis looking bright. Rojo looking like Rojo.
Output: joy, optimism

Input: Yet again another night I should've stayed in😊
Output: joy

Input: I hope a goal comes soon on either side. Otherwise there is a serious threat of a dull and frustrating game. #COYS
Output: anger, disgust, pessimism, sadness

Input: Watch this amazing live.ly broadcast by @rosannahill  #musically
Output: joy, surprise

Now I want you to label the following example:
Input: Will be a lively evening at ot that one.
Output: 
```

### CODE REVIEW

File: tests/assets/sem_eval_2018_task_1/seed.csv [hunk 1]
```diff
@@ -0,0 +1,70 @@
+example,label
```
[the remark this answers is not in the document]
Comment (rishabh-bhargava) [reply tests/assets/sem_eval_2018_task_1/seed.csv:1]: also, from what I can tell, the `seed.csv` and `test.csv` files that are being checked into the repo are much smaller. take a look at the `assets/` folder in `test/`?

**Review by rishabh-bhargava — COMMENTED** [review]
just added some minor comments. haven't gone through the code too deeply

File: tests/assets/sem_eval_2018_task_1/seed.csv (hunk 1 above)
[the remark this answers is not in the document]
Comment (nihit) [reply tests/assets/sem_eval_2018_task_1/seed.csv:1]: Yes, these should be just dummy seed and test files for testing (not the entire seed and test sets)

Additionally, @Tyrest  I see ~6700 rows in test and only ~70 rows in seed. This doesn't matter for the unit testing assets but want to make sure that the actual dataset files being checked into https://s3.console.aws.amazon.com/s3/buckets/autolabel-benchmarking are 200 labeled examples for seed, and 2000 for test for the benchmarking

File: tests/assets/sem_eval_2018_task_1/config_sem_eval_2018_task_1.json [hunk 2]
```diff
@@ -0,0 +1,34 @@
+{
```
Comment (nihit) [inline tests/assets/sem_eval_2018_task_1/config_sem_eval_2018_task_1.json:1]: let's rename this file to twitter_emotion_detection to avoid confusion

File: tests/unit/tasks_test.py [hunk 3]
```diff
@@ -19,6 +20,10 @@
 
 SCIQ_CONFIG = json.load(open("tests/assets/sciq/config_sciq.json", "r"))
 
+SEM_EVAL_2018_CONFIG = json.load(
```
Comment (nihit) [inline tests/unit/tasks_test.py:23]: same - rename to twitter emotion detection config

File: examples/twitter_emotion_detection/config_twitter_emotion_detection.json [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    "delimiter": ","
+  },
+  "model": {
+    "provider": "openai",
+    "name": "gpt-3.5-turbo"
+  },
+  "prompt": {
+    "task_guidelines": "You are an expert at classifying tweets as neutral or one or more of the given emotions that best represent the mental state of the poster.\nYour job is to correctly label the provided input example into one or more of the following categories:\n{labels}",
+    "output_guidelines": "You will return the answer as a comma separated list of labels sorted in alphabetical order. For example: \"label1, label2, label3\"",
+    "labels": [
```
[the remark this answers is not in the document]
Comment (nihit) [reply examples/twitter_emotion_detection/config_twitter_emotion_detection.json:15]: @rishabh-bhargava i think it's reasonable to expect labeling guidelines to take care of it?

File: tests/assets/sem_eval_2018_task_1/seed.csv (hunk 1 above)
[the remark this answers is not in the document]
Comment (Tyrest) [reply tests/assets/sem_eval_2018_task_1/seed.csv:1]: Yep, didn't realize that the `csv` files were smaller. I'll make these changes now!

File: src/autolabel/tasks/multi_label_classification.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+from autolabel.schema import LLMAnnotation, Metric, MetricResult
+from autolabel.tasks import BaseTask
+from autolabel.tasks.utils import compute_f1
+from autolabel.utils import get_format_variables
+
+import json
+
+
+class MultiLabelClassificationTask(BaseTask):
+    DEFAULT_OUTPUT_GUIDELINES = 'You will return the answer as a comma separated list of labels. For example: "label1, label2, label3"'
```
Comment (nihit) [inline src/autolabel/tasks/multi_label_classification.py:18]: how well does this guideline do across LLMs ? A few other choices possible here are:
- return it separated by newlines
- return it as a list/array

I'm neutral to what we choose ultimately, but whatever is the default output guideline should work well across LLMs (including Refuel LLM)

File: src/autolabel/tasks/multi_label_classification.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+        for index, l in enumerate(llm_labels):
+            if l.label != self.NULL_LABEL_TOKEN and (
+                l.confidence_score is None or l.confidence_score >= threshold
+            ):
+                answered_llm_preds.append(l.label.lower())
+                answered_gt_labels.append(gt_labels[index].lower())
+
+        return answered_gt_labels, answered_llm_preds
+
+    def eval(
```
Comment (nihit) [inline src/autolabel/tasks/multi_label_classification.py:118]: @DhruvaBansal00  can you share some resources/pointers for multi-label eval metrics that would be good to add here?

File: src/autolabel/tasks/multi_label_classification.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+        self, llm_labels: List[LLMAnnotation], gt_labels: List[str]
+    ) -> List[MetricResult]:
+        """Evaluate the LLM generated labels by comparing them against ground truth
+
+        Args:
+            llm_labels (List[LLMAnnotation]): _description_
+            gt_labels (List[str]): _description_
```
Comment (nihit) [inline src/autolabel/tasks/multi_label_classification.py:125]: add description please

**Comment by nihit** [comment]
issue: https://github.com/refuel-ai/autolabel/issues/391

File: src/autolabel/tasks/multilabel_classification.py [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+        )
+
+    def auroc_score_labels(
+        self, gt_labels, llm_labels
+    ) -> Tuple[List[int], List[float]]:
+        labels = []
+        confidences = []
+        for index, llm_label in enumerate(llm_labels):
+            labels.append(llm_label.label.lower() == gt_labels[index].lower())
+            confidences.append(llm_label.confidence_score)
```
Comment (DhruvaBansal00) [inline src/autolabel/tasks/multilabel_classification.py:103]: Confidence here will be the average of log probs of all the tokens in the llm labels. 

We might instead want to define confidence individually for each each label in the list. Not sure if that is in the scope of this PR - if not, let's open a ticket for doing this as a follow up.

Comment (DhruvaBansal00) [reply src/autolabel/tasks/multilabel_classification.py:103]: Implementing this will require some revamp on the confidence calculation side. We probably want to now move confidence calculation into each task (and the parent class can implement the default confidence calculator)

File: src/autolabel/tasks/multilabel_classification.py [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+from typing import List, Dict, Tuple
+
+from langchain.prompts.prompt import PromptTemplate
+from sklearn.metrics import accuracy_score
+
+from autolabel.confidence import ConfidenceCalculator
+from autolabel.configs import AutolabelConfig
+from autolabel.schema import LLMAnnotation, Metric, MetricResult
+from autolabel.tasks import BaseTask
+from autolabel.tasks.utils import compute_f1
```
Comment (DhruvaBansal00) [inline src/autolabel/tasks/multilabel_classification.py:11]: I think compute_f1 requires some changes as a part of this PR: https://github.com/refuel-ai/autolabel/blob/107962437ab04f024b7942380b965a79f94a4f59/src/autolabel/tasks/utils.py#L25

That function currently splits string using .split() and the default separator. We can either change that function to take in a list of tokens that are pre-split or allow sending a custom separator. I believe this classes uses ',' as the separator between labels.

Comment (DhruvaBansal00) [reply src/autolabel/tasks/multilabel_classification.py:11]: Also, there we are calculating a micro F1 score. We can instead use https://scikit-learn.org/stable/modules/generated/sklearn.metrics.f1_score.html to give the user the option to change between micro, macro, weighted, or class wise F1 scores.

File: src/autolabel/tasks/multilabel_classification.py (hunk 8 above)
Comment (Tyrest) [reply src/autolabel/tasks/multilabel_classification.py:103]: Just created #406 for this

File: src/autolabel/tasks/multilabel_classification.py (hunk 9 above)
Comment (Tyrest) [reply src/autolabel/tasks/multilabel_classification.py:11]: Since `QuestionAnsweringTask` uses this method as well, I'm just going to leave `compute_f1` as is for now and use `sklearn`'s `f1_score` directly in `MultilabelClassificationTask.eval`

**Review by nihit — APPROVED** [review]

**Comment by nihit** [comment]
@DhruvaBansal00 will let you take a final pass, lgtm

**Review by DhruvaBansal00 — APPROVED** [review]
Lgtm!

However, I do think we should consolidate f1 calculation for multilabel and question answering into one function. We should be using sklearn to calculate it for both. Let's open a ticket to track that?
